src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java

```diff
@@ -0,0 +1,439 @@
+        var front = isPowered() ?
+                Textures.QUANTUM_CONTROLLER_FRONT_ACTIVE :
+                Textures.QUANTUM_CONTROLLER_FRONT_INACTIVE;
+        var sides = isPowered() ?
+                Textures.QUANTUM_CONTROLLER_ACTIVE :
+                Textures.QUANTUM_CONTROLLER_INACTIVE;
+
+        var newPipeline = ArrayUtils.add(pipeline,
+                new ColourMultiplier(GTUtility.convertRGBtoOpaqueRGBA_CL(getPaintingColorForRendering())));
+
+        front.renderSided(getFrontFacing(), renderState, translation, newPipeline);
+
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            if (facing == getFrontFacing()) continue;
+            sides.renderSided(facing, renderState, translation, newPipeline);
+        }
+    }
+
+    @Override
+    public Pair<TextureAtlasSprite, Integer> getParticleTexture() {
+        return Pair.of(Textures.QUANTUM_CONTROLLER_ACTIVE.getParticleSprite(), getPaintingColorForRendering());
+    }
+
+    @Override
+    protected boolean openGUIOnRightClick() {
+        return false; // todo use mui2 for ui?
+    }
+
+    @Override
+    public boolean isValidFrontFacing(EnumFacing facing) {
+        return true;
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private IQuantumStorage<?> getStorage(BlockPos pos, boolean rebuild) {
+        if (getWorld().isRemote) return null;
+        if (storageInstances.containsKey(pos)) {
+            WeakReference<IQuantumStorage<?>> storageRef = storageInstances.get(pos);
+            IQuantumStorage<?> storage = storageRef.get();
+            if (storage != null) {
+                return storage;
+            }
+        }
+        // need to make sure it still exists
+        MetaTileEntity mte = GTUtility.getMetaTileEntity(getWorld(), pos);
+        if (mte instanceof IQuantumStorage<?>storage) {
+            storageInstances.put(pos, new WeakReference<>(storage));
+            return storage;
+        } else if (rebuild) {
+            // need to remove and rebuild
+            storagePositions.remove(pos);
+            rebuildNetwork();
+            return null;
+        }
+        return null;
+    }
+
+    @Override
+    public boolean canConnect(IQuantumStorage<?> storage) {
+        return !isDead && isInRange(storage.getPos());
+    }
+
+    private boolean isInRange(BlockPos pos) {
+        boolean isXValid = Math.abs(getPos().getX() - pos.getX()) <= MAX_DISTANCE_RADIUS;
+        boolean isYValid = Math.abs(getPos().getY() - pos.getY()) <= MAX_DISTANCE_RADIUS;
+        boolean isZValid = Math.abs(getPos().getZ() - pos.getZ()) <= MAX_DISTANCE_RADIUS;
+        return isXValid && isYValid && isZValid;
+    }
+
+    @Override
+    public void onRemoval() {
+        if (getWorld().isRemote) return;
+        isDead = true;
+        for (BlockPos pos : storagePositions) {
+            IQuantumStorage<?> storage = getStorage(pos, false);
+            if (storage != null) storage.setDisconnected();
```

Comment:
This call could get a bit expensive, given that it is a loop over all the valid storages in the network, and each call to disconnect the storage sees the storage attempt to find a new network. Do you have any ideas about how to potentially improve this, or do you see it not being an issue, even on large networks?

Reply:
there might be a case where a quantum part could connect to a new valid controller

Reply:
actually moved the network checking out, so it's better now